From a655a2ab54d3e0bbfdb771d5ce7dfe3570724170 Mon Sep 17 00:00:00 2001
From: Lars Ekman <lars.g.ekman@est.tech>
Date: Sat, 5 Dec 2020 08:32:15 +0100
Subject: [PATCH] Ipvs: non-local access to externalTrafficPolicy:Local

Allow access to externalTrafficPolicy:Local services from PODs
not on a node where a server executes. Problem described in #93456
---
 pkg/proxy/ipvs/proxier.go | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/pkg/proxy/ipvs/proxier.go b/pkg/proxy/ipvs/proxier.go
index 7d67008332c3c..3cc271a201e32 100644
--- a/pkg/proxy/ipvs/proxier.go
+++ b/pkg/proxy/ipvs/proxier.go
@@ -2046,6 +2046,12 @@ func (proxier *Proxier) syncEndpoint(svcPortName proxy.ServicePortName, onlyNode
 		newEndpoints.Insert(epInfo.String())
 	}
 
+	if len(newEndpoints) == 0 && onlyNodeLocalEndpoints {
+		for _, epInfo := range proxier.endpointsMap[svcPortName] {
+			newEndpoints.Insert(epInfo.String())
+		}
+	}
+
 	// Create new endpoints
 	for _, ep := range newEndpoints.List() {
 		ip, port, err := net.SplitHostPort(ep)
